go-controller/pkg/node/gateway_localnet.go

```diff
@@ -239,88 +246,357 @@ func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
+				klog.Errorf("error delete routing table entry for ExternalIP %s: stdout: %s, stderr: %s, err: %v", externalIP, stdout, stderr, err)
+			}
+		}
 	}
-	rules := localnetIptRules(svc, npw.gatewayIP)
-	klog.V(5).Infof("Delete rules %v for service %v", rules, svc.Name)
-	delIptRules(npw.ipt, rules)
-	return nil
+	klog.V(5).Infof("delete iptables rules: %v for service: %v", iptRules, svc.Name)
+	return delIptRules(npw.ipt, iptRules)
 }
 
-func localnetNodePortWatcher(ipt util.IPTablesHelper, wf *factory.WatchFactory, gatewayIP net.IP) error {
-	// delete all the existing OVN-NODEPORT rules
-	// TODO: Add a localnetSyncService method to remove the stale entries only
-	_ = ipt.ClearChain("nat", iptableNodePortChain)
-	_ = ipt.ClearChain("filter", iptableNodePortChain)
+func (npw *localnetNodePortWatcherData) syncServices(serviceInterface []interface{}) {
+	removeStaleIPTRules := func(table, chain string, serviceRules []iptRule) {
+		if ipTableRules, err := npw.ipt.List(table, chain); err != nil {
+			for _, ipTableRule := range ipTableRules {
+				isFound := false
+				for _, serviceRule := range serviceRules {
+					if ipTableRule == strings.Join(serviceRule.args, " ") {
+						isFound = true
+						break
+					}
+				}
+				if !isFound {
+					klog.V(5).Infof("deleting stale iptables rule: table: %s, chain: %s, rule: %s", table, chain, ipTableRule)
+					if err := npw.ipt.Delete(table, chain, strings.Fields(ipTableRule)...); err != nil {
+						klog.Errorf("error deleting stale iptables rule: table: %s, chain: %s, rule: %v", table, chain, ipTableRule)
+					}
+				}
+			}
+		}
+	}
+	removeStaleRoutes := func(externalIPs []string) {
+		stdout, stderr, err := util.RunIP("route", "list", "table", localnetGatwayExternalIDTable)
+		if err != nil || stdout == "" {
+			klog.Infof("no routing table entries for ExternalIP table %s: stdout: %s, stderr: %s, err: %v", localnetGatwayExternalIDTable, stdout, stderr, err)
+			return
+		}
+		for _, existingRoute := range strings.Split(stdout, "\n") {
+			isFound := false
+			for _, externalIP := range externalIPs {
+				if strings.Contains(existingRoute, externalIP) {
+					isFound = true
+					break
+				}
+			}
+			if !isFound {
+				klog.V(5).Infof("deleting stale routing rule: %s", existingRoute)
+				if _, stderr, err := util.RunIP("route", "del", existingRoute); err != nil {
+					klog.Errorf("error deleting stale routing rule: stderr: %s, err: %v", stderr, err)
+				}
+			}
+		}
+	}
 
-	rules := make([]iptRule, 0)
-	rules = append(rules, iptRule{
-		table: "nat",
-		chain: "PREROUTING",
-		args:  []string{"-j", iptableNodePortChain},
-	})
-	rules = append(rules, iptRule{
-		table: "nat",
-		chain: "OUTPUT",
-		args:  []string{"-j", iptableNodePortChain},
-	})
-	rules = append(rules, iptRule{
-		table: "filter",
-		chain: "FORWARD",
-		args:  []string{"-j", iptableNodePortChain},
-	})
+	// Here it's actually in our favour to skip the conditions listed in addService
```

Comment:
I don't think these comments add anything other than confusion. I'm not sure how we are increasing probability here or doing anything different than how I would expect a delete sync to occur.

Reply:
Ok, I will remove these comments. Just FYI on what I meant by "increases the probability": by not having the additional checks that we do have in `addService`/ `deleteService` we "throw a larger net and capture" data that might have gotten added incorrectly previously, and clean that up.